execution/gethexec/stylus_tracer.go

```diff
@@ -0,0 +1,138 @@
+// Copyright 2024, Offchain Labs, Inc.
+// For license information, see https://github.com/OffchainLabs/nitro/blob/master/LICENSE
+
+package gethexec
+
+import (
+	"encoding/json"
+	"math/big"
+	"sync/atomic"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/core/vm"
+	"github.com/ethereum/go-ethereum/eth/tracers"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+func init() {
+	tracers.DefaultDirectory.Register("stylusTracer", newStylusTracer, false)
+}
+
+// stylusTracer captures Stylus HostIOs and returns them in a structured format to be used in Cargo
+// Stylus Replay.
+type stylusTracer struct {
+	open      *[]HostioTraceInfo
+	stack     []*[]HostioTraceInfo
+	interrupt atomic.Bool
+	reason    error
+}
+
+// HostioTraceInfo contains the captured HostIO log returned by stylusTracer.
+type HostioTraceInfo struct {
+	Name     string             `json:"name"`
+	Args     hexutil.Bytes      `json:"args"`
+	Outs     hexutil.Bytes      `json:"outs"`
+	StartInk uint64             `json:"startInk"`
+	EndInk   uint64             `json:"endInk"`
+	Address  *common.Address    `json:"address,omitempty"`
+	Steps    *[]HostioTraceInfo `json:"steps,omitempty"`
+}
+
+// nestsHostios contains the hostios with nested calls.
+var nestsHostios = map[string]bool{
+	"call_contract":          true,
+	"delegate_call_contract": true,
+	"static_call_contract":   true,
+}
+
+func newStylusTracer(ctx *tracers.Context, _ json.RawMessage) (tracers.Tracer, error) {
+	var open []HostioTraceInfo
+	return &stylusTracer{
+		open: &open,
+	}, nil
+}
+
+func (t *stylusTracer) CaptureStylusHostio(name string, args, outs []byte, startInk, endInk uint64) {
+	if t.interrupt.Load() {
+		return
+	}
+	info := HostioTraceInfo{
+		Name:     name,
+		Args:     args,
+		Outs:     outs,
+		StartInk: startInk,
+		EndInk:   endInk,
+	}
+	if nestsHostios[name] {
+		last := pop(t.open)
+		info.Address = last.Address
+		info.Steps = last.Steps
+	}
+	*t.open = append(*t.open, info)
+}
+
+func (t *stylusTracer) CaptureEnter(typ vm.OpCode, from common.Address, to common.Address, input []byte, gas uint64, value *big.Int) {
+	if t.interrupt.Load() {
+		return
+	}
+	inner := []HostioTraceInfo{}
+	info := HostioTraceInfo{
+		Address: &to,
+		Steps:   &inner,
+	}
```

Comment:
oh, your unit test will tell you what's needed once you add the right testcase :)

Reply:
I don't like setting the name to type because this is the opcode name, but the trace returns the HostIO name. The stylus replay will only work on the top-level stylus contract, so adding more info on these inner EVM calls is not helpful for the tool. Instead, the trace could have a field called `isEvm` to make it clear when it enters an EVM contract. What do you think?

Reply:
We agreed to add the `evm_` prefix to the name of evm calls.